src/MonoTorrent/MonoTorrent/TorrentCreator.cs

```diff
@@ -180,62 +181,108 @@ void AddCommonStuff (BEncodedDictionary torrent)
             torrent ["creation date"] = new BEncodedNumber ((long) span.TotalSeconds);
         }
 
-        byte [] CalcPiecesHash (List<TorrentFile> files, IPieceWriter writer, CancellationToken token)
+        async Task<byte []> CalcPiecesHashAsync (List<TorrentFile> files, IPieceWriter writer, CancellationToken token)
         {
-            int bufferRead = 0;
-            long fileRead = 0;
-            long overallRead = 0;
-            MD5 md5Hasher = null;
-
-            var shaHasher = HashAlgoFactory.Create<SHA1> ();
             var torrentHashes = new List<byte> ();
             var overallTotal = files.Sum (t => t.Length);
-            var buffer = new byte [PieceLength];
 
-            if (StoreMD5)
-                md5Hasher = HashAlgoFactory.Create<MD5> ();
+            int parallelTasks = Environment.ProcessorCount;
+            var emptyBuffers = new AsyncProducerConsumerQueue<byte[]> (parallelTasks + 1);
+            var filledBuffers = new AsyncProducerConsumerQueue<(byte[], int, TorrentFile)> (parallelTasks);
 
-            try {
-                foreach (TorrentFile file in files) {
-                    fileRead = 0;
-                    md5Hasher?.Initialize ();
+            for (int i = 0; i < parallelTasks; i ++)
+                await emptyBuffers.EnqueueAsync (new byte[BlockSize]);
```

Comment:
You could use ArrayPool here.

Reply:
Yup, it's definitely a good candidate for buffer re-use, and also a good candidate for `ArrayPool`.

`ArrayPool` isn't built into .NET Standard 2.0, though it is available as a NuGet. I'm a little dependency averse so i've been resisting the urge to delete [pre-existing implementations](https://github.com/mono/monotorrent/blob/master/src/MonoTorrent/MonoTorrent.Client/Managers/BufferManager.cs) and replace them with NuGets.

Apart from that, I decided to not use pooled buffers while I work the underlying changes. It should be relatively easy to swap to pooled buffers when this is done :)